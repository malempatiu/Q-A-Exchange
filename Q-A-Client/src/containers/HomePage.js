import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchQuestions, removeError} from '../actions/actions';
import { Link } from 'react-router-dom';

class HomePage extends Component {
    state = {
        text: ''
    }
    componentDidMount() {
        this.props.removeError();
        this.props.fetchQuestions();
    };
    handleUserSearch = (e) => {
        this.setState({ text: e.target.value });
    };
    render() {
        const { questions, error } = this.props;
        let newQuestions;
        if (this.state.text.length > 0) {
            newQuestions = questions.filter((question) => {
                return question.text.toLowerCase().includes(this.state.text.toLowerCase());
            });
        } else {
            newQuestions = questions;
        }
        return (
            <div className="conatiner homepage">
                <div className="jumbotron text-center">
                    <h1>Welcome to Q-A-Exchange</h1>
                    <p className="lead">View all top questions and answers</p>
                    <Link className="btn btn-info text-white" to="/addquestion">Ask Question</Link>
                </div>
                <div className="container text-center w-75">
                    <input type="text" className="form-control searchInput" placeholder="search questions....." onChange={this.handleUserSearch} style={{ margin: '8px' }} />
                </div>
                <div className="container">
                    {error && <div className="alert alert-danger">{error}</div>}
                    {newQuestions.map((question) => {
                        return <div key={question._id} className="row">
                            <div className="col-lg-12 col-md-12 col-sm-12">
                                <div className="card">
                                    <Link to={`/question/${question._id}`} id={question._id} className="card-body text-primary">
                                        {question.text}
                                    </Link>
                                </div>
                            </div>
                        </div>
                    })}
                </div>
            </div>
        );
    }
};

const mapStateToProps = (state) => {
    return {
        questions: state.Questions.questions,
        error: state.Errors.error
    };
}

export default connect(mapStateToProps, { fetchQuestions, removeError })(HomePage);