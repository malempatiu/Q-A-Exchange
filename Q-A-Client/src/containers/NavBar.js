import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {logoutUser} from '../actions/actions';


export class NavBar extends Component {
    handleLogout = (e) => {
        this.props.logoutUser();       
    };
    render() {
        const { isAuthenticated, username } = this.props;
        return (
            <nav className="navbar navbar-expand-lg navbar-dark bg-dark sticky-top">
                <div className="container">
                    <h1 className="navbar-brand text-white font-weight-bold">Q-A-Exchange</h1>
                    <Link className="nav-link" to="/">Home</Link>
                    <div className="navbar-toggler" data-toggle="collapse" data-target="#navbarMenu">
                        <span className="navbar-toggler-icon"></span>
                    </div>
                    <div className="collapse navbar-collapse" id="navbarMenu">
                        <div className="navbar-nav ml-auto">
                            {isAuthenticated ?
                                <div>
                                    <a className="nav-item nav-link text-success font-weight-bold" to="/">welcome {username}</a>
                                    <Link className="nav-item nav-link btn btn-info text-white" to="/userquestions">My Questions</Link>
                                    <Link className="nav-item nav-link btn btn-info text-white" to="/" onClick={this.handleLogout}>Log Out</Link>
                                </div>
                                :
                                <div>
                                    <Link className="nav-item nav-link btn btn-info text-white" to="/signup">Sign Up</Link>
                                    <Link className="nav-item nav-link btn btn-info text-white" to="/signin">Log In</Link>
                                </div>
                            }
                        </div>
                    </div>
                </div>
            </nav>
        );
    };
};

const mapStateToProps = (state) => {
    return {
        isAuthenticated: state.CurrentUser.isAuthenticated,
        username: state.CurrentUser.user.username
    };
}

export default connect(mapStateToProps, {logoutUser})(NavBar);