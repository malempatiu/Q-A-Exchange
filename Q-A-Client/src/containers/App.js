import React, { Component } from 'react';
import { BrowserRouter, Switch, Route } from 'react-router-dom';
import { combineReducers, createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Questions from '../reducers/questions-reducer';
import Errors from '../reducers/errors-reducer';
import QuestionAndAnswer from '../reducers/question-answer-reducer';
import NavBar from './NavBar';
import HomePage from './HomePage';
import CreateQuestion from '../components/CreateQuestion';
import QuestionAnswer from '../components/QuestionAnswer';

const store = createStore(
  combineReducers({
    Questions,
    QuestionAndAnswer,
    Errors
  }),
  compose(applyMiddleware(thunk))
);

class App extends Component {
  render() {
    return (
      <Provider store={store}>
        <div className="app-cont">
          <BrowserRouter>
            <div>
              <NavBar />
              <Switch>
                <Route exact path="/" component={HomePage} />
                <Route exact path="/addquestion" component={CreateQuestion} />
                <Route exact path="/question/:id" component={QuestionAnswer} />
              </Switch>
            </div>
          </BrowserRouter>
        </div>
      </Provider>
    );
  }
}

export default App;

