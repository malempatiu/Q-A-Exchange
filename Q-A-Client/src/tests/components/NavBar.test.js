import React from 'react';
import { NavBar } from '../../containers/NavBar';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('NAVBAR CONTAINER', () => {
    it('should render navbar with home, signup & login links', () => {
        const wrapper = shallow(<NavBar />);
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('Link')).toHaveLength(3);
    });
    it('should render navbar with home, myquestions & logout links', () => {
        const wrapper = shallow(<NavBar isAuthenticated={true} username={'ugesh'} />);
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('a')).toHaveLength(1);
        expect(wrapper.find('Link')).toHaveLength(3);
    });
    it('should call handleLogout', () => {
         const logoutUserSpy = jest.fn();
         const wrapper = shallow(<NavBar isAuthenticated={true} username={'ugesh'} logoutUser={logoutUserSpy}/>);
         wrapper.find('Link').at(2).simulate('click');
         expect(logoutUserSpy).toHaveBeenCalled();
    });
});