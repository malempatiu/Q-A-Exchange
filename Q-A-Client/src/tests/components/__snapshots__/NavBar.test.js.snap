// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NAVBAR CONTAINER should render navbar with home, myquestions & logout links 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark sticky-top"
>
  <div
    className="container"
  >
    <h1
      className="navbar-brand text-white font-weight-bold"
    >
      Q-A-Exchange
    </h1>
    <Link
      className="nav-link"
      replace={false}
      to="/"
    >
      Home
    </Link>
    <div
      className="navbar-toggler"
      data-target="#navbarMenu"
      data-toggle="collapse"
    >
      <span
        className="navbar-toggler-icon"
      />
    </div>
    <div
      className="collapse navbar-collapse"
      id="navbarMenu"
    >
      <div
        className="navbar-nav ml-auto"
      >
        <div>
          <a
            className="nav-item nav-link text-success font-weight-bold"
            to="/"
          >
            welcome 
            ugesh
          </a>
          <Link
            className="nav-item nav-link btn btn-info text-white"
            replace={false}
            to="/userquestions"
          >
            My Questions
          </Link>
          <Link
            className="nav-item nav-link btn btn-info text-white"
            onClick={[Function]}
            replace={false}
            to="/"
          >
            Log Out
          </Link>
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`NAVBAR CONTAINER should render navbar with home, signup & login links 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark sticky-top"
>
  <div
    className="container"
  >
    <h1
      className="navbar-brand text-white font-weight-bold"
    >
      Q-A-Exchange
    </h1>
    <Link
      className="nav-link"
      replace={false}
      to="/"
    >
      Home
    </Link>
    <div
      className="navbar-toggler"
      data-target="#navbarMenu"
      data-toggle="collapse"
    >
      <span
        className="navbar-toggler-icon"
      />
    </div>
    <div
      className="collapse navbar-collapse"
      id="navbarMenu"
    >
      <div
        className="navbar-nav ml-auto"
      >
        <div>
          <Link
            className="nav-item nav-link btn btn-info text-white"
            replace={false}
            to="/signup"
          >
            Sign Up
          </Link>
          <Link
            className="nav-item nav-link btn btn-info text-white"
            replace={false}
            to="/signin"
          >
            Log In
          </Link>
        </div>
      </div>
    </div>
  </div>
</nav>
`;
